mage_ai/api/resources/PipelineResource.py

```diff
@@ -414,6 +418,40 @@ def retry_pipeline_runs(pipeline_runs):
             for run in pipeline_runs:
                 retry_pipeline_run(run)
 
+        @safe_db_query
+        def query_incomplete_block_runs(pipeline_uuid: str):
+            a = aliased(PipelineRun, name='a')
+            b = aliased(BlockRun, name='b')
+            columns = [
+                b.id,
+                b.status,
+                b.pipeline_run_id,
+                a.status,
+                a.pipeline_uuid,
+            ]
+            result = (
+                PipelineRun.
+                select(*columns).
+                join(b, a.id == b.pipeline_run_id).
+                filter(a.pipeline_uuid == pipeline_uuid).
+                filter(b.status != BlockRun.BlockRunStatus.COMPLETED)
```

Comment:
Also filter out `BlockRun.BlockRunStatus.CONDITION_FAILED` status.

Reply:
Let's only retry failed pipeline runs.

Reply:
@wangxiaoyou1993 Made some updates on both BE and FE so only incomplete block runs for failed pipeline runs will be retried, meaning if a pipeline run was manually canceled, its block runs would not be retried.

![retry block runs only for failed pipeline runs](https://github.com/mage-ai/mage-ai/assets/78053898/23a869dd-e960-4df3-83e5-06fc06c2217c)